### ISSUE TITLE
Test failure in runtest.py::test_multithreading.

### ISSUE THREAD

**Issue by robertnishihara** [ray-project/ray#3754]
The test
```
python -m pytest -v test/runtest.py::test_multithreading
```
is failing on the current master, e.g., [1]. The output is copied below. It looks to me like the same object ID is being generated multiple times in the test and so the second attempt to get the object actually retrieves the first value that was created.

```
_____________________________ test_multithreading ______________________________
shutdown_only = None
    def test_multithreading(shutdown_only):
        # This test requires at least 2 CPUs to finish since the worker does not
        # relase resources when joining the threads.
        ray.init(num_cpus=2)
    
        def run_test_in_multi_threads(test_case, num_threads=20, num_repeats=50):
            """A helper function that runs test cases in multiple threads."""
    
            def wrapper():
                for _ in range(num_repeats):
                    test_case()
                    time.sleep(random.randint(0, 10) / 1000.0)
                return "ok"
    
            executor = ThreadPoolExecutor(max_workers=num_threads)
            futures = [executor.submit(wrapper) for _ in range(num_threads)]
            for future in futures:
                assert future.result() == "ok"
    
        @ray.remote
        def echo(value, delay_ms=0):
            if delay_ms > 0:
                time.sleep(delay_ms / 1000.0)
            return value
    
        @ray.remote
        class Echo(object):
            def echo(self, value):
                return value
    
        def test_api_in_multi_threads():
            """Test using Ray api in multiple threads."""
    
            # Test calling remote functions in multiple threads.
            def test_remote_call():
                value = random.randint(0, 1000000)
                result = ray.get(echo.remote(value))
                assert value == result
    
            run_test_in_multi_threads(test_remote_call)
    
            # Test multiple threads calling one actor.
            actor = Echo.remote()
    
            def test_call_actor():
                value = random.randint(0, 1000000)
                result = ray.get(actor.echo.remote(value))
                assert value == result
    
            run_test_in_multi_threads(test_call_actor)
    
            # Test put and get.
            def test_put_and_get():
                value = random.randint(0, 1000000)
                result = ray.get(ray.put(value))
                assert value == result
    
            run_test_in_multi_threads(test_put_and_get)
    
            # Test multiple threads waiting for objects.
            num_wait_objects = 10
            objects = [
                echo.remote(i, delay_ms=10) for i in range(num_wait_objects)
            ]
    
            def test_wait():
                ready, _ = ray.wait(
                    objects,
                    num_returns=len(objects),
                    timeout=1000,
                )
                assert len(ready) == num_wait_objects
                assert ray.get(ready) == list(range(num_wait_objects))
    
            run_test_in_multi_threads(test_wait, num_repeats=1)
    
        # Run tests in a driver.
        test_api_in_multi_threads()
    
        # Run tests in a worker.
        @ray.remote
        def run_tests_in_worker():
            test_api_in_multi_threads()
            return "ok"
    
>       assert ray.get(run_tests_in_worker.remote()) == "ok"
test/runtest.py:1261: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
object_ids = ObjectID(0100000019b8772154e0349396183a20684a3444)
worker = <ray.worker.Worker object at 0x7f81caef4f10>
    def get(object_ids, worker=global_worker):
        """Get a remote object or a list of remote objects from the object store.
    
        This method blocks until the object corresponding to the object ID is
        available in the local object store. If this object is not in the local
        object store, it will be shipped from an object store that has it (once the
        object has been created). If object_ids is a list, then the objects
        corresponding to each object in the list will be returned.
    
        Args:
            object_ids: Object ID of the object to get or a list of object IDs to
                get.
    
        Returns:
            A Python object or a list of Python objects.
    
        Raises:
            Exception: An exception is raised if the task that created the object
                or that created one of the objects raised an exception.
        """
        worker.check_connected()
        with profiling.profile("ray.get", worker=worker):
            if worker.mode == LOCAL_MODE:
                # In LOCAL_MODE, ray.get is the identity operation (the input will
                # actually be a value not an objectid).
                return object_ids
            global last_task_error_raise_time
            if isinstance(object_ids, list):
                values = worker.get_object(object_ids)
                for i, value in enumerate(values):
                    if isinstance(value, RayTaskError):
                        last_task_error_raise_time = time.time()
                        raise value
                return values
            else:
                value = worker.get_object([object_ids])[0]
                if isinstance(value, RayTaskError):
                    # If the result is a RayTaskError, then the task that created
                    # this object failed, and we should propagate the error message
                    # here.
                    last_task_error_raise_time = time.time()
>                   raise value
E                   RayTaskError: ray_worker:runtest.run_tests_in_worker() (pid=26925, host=travis-job-b734bee9-bcb9-4172-91db-7f33f457d6b1)
E                     File "/home/travis/build/ray-project/ray/test/runtest.py", line 1258, in run_tests_in_worker
E                       test_api_in_multi_threads()
E                     File "/home/travis/build/ray-project/ray/test/runtest.py", line 1233, in test_api_in_multi_threads
E                       run_test_in_multi_threads(test_put_and_get)
E                     File "/home/travis/build/ray-project/ray/test/runtest.py", line 1193, in run_test_in_multi_threads
E                       assert future.result() == "ok"
E                     File "/home/travis/miniconda/lib/python2.7/site-packages/concurrent/futures/_base.py", line 455, in result
E                       return self.__get_result()
E                     File "/home/travis/miniconda/lib/python2.7/site-packages/concurrent/futures/thread.py", line 63, in run
E                       result = self.fn(*self.args, **self.kwargs)
E                     File "/home/travis/build/ray-project/ray/test/runtest.py", line 1186, in wrapper
E                       test_case()
E                     File "/home/travis/build/ray-project/ray/test/runtest.py", line 1231, in test_put_and_get
E                       assert value == result
E                   AssertionError: assert 243799 == 900664
../../../.local/lib/python2.7/site-packages/ray-0.6.1-py2.7-linux-x86_64.egg/ray/worker.py:2254: RayTaskError
----------------------------- Captured stderr call -----------------------------
No handlers could be found for logger "ray.memory_monitor"
No handlers could be found for logger "ray.memory_monitor"
No handlers could be found for logger "ray.memory_monitor"
No handlers could be found for logger "ray.memory_monitor"
Calling ray.get or ray.wait in a separate thread may lead to deadlock if the main thread blocks on this thread and there are not enough resources to execute more tasks
Exception in thread ray_import_thread (most likely raised during interpreter shutdown):
Traceback (most recent call last):
  File "/home/travis/miniconda/lib/python2.7/threading.py", line 801, in __bootstrap_inner
  File "/home/travis/miniconda/lib/python2.7/threading.py", line 754, in run
  File "/home/travis/.local/lib/python2.7/site-packages/ray-0.6.1-py2.7-linux-x86_64.egg/ray/import_thread.py", line 71, in _run
<type 'exceptions.AttributeError'>: 'NoneType' object has no attribute 'ConnectionError'
The object with ID ObjectID(ffffffff4d97776d1fd6a47ce4a495cbee8fc012) already exists in the object store.
Pushing error to dirver, type: task, message: ray_worker:runtest.run_tests_in_worker() (pid=26925, host=travis-job-b734bee9-bcb9-4172-91db-7f33f457d6b1)
  File "/home/travis/build/ray-project/ray/test/runtest.py", line 1258, in run_tests_in_worker
    test_api_in_multi_threads()
  File "/home/travis/build/ray-project/ray/test/runtest.py", line 1233, in test_api_in_multi_threads
    run_test_in_multi_threads(test_put_and_get)
  File "/home/travis/build/ray-project/ray/test/runtest.py", line 1193, in run_test_in_multi_threads
    assert future.result() == "ok"
  File "/home/travis/miniconda/lib/python2.7/site-packages/concurrent/futures/_base.py", line 455, in result
    return self.__get_result()
  File "/home/travis/miniconda/lib/python2.7/site-packages/concurrent/futures/thread.py", line 63, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/travis/build/ray-project/ray/test/runtest.py", line 1186, in wrapper
    test_case()
  File "/home/travis/build/ray-project/ray/test/runtest.py", line 1231, in test_put_and_get
    assert value == result
AssertionError: assert 243799 == 900664
```

[1] https://travis-ci.com/ray-project/ray/jobs/169722588

**Comment by raulchen** [comment]
I couldn't reproduce this on my macbook. Will find a linux machine to test this tomorrow.

**Comment by raulchen** [comment]
I'm able to reproduce this now. The chance of failure is about 10%. But I still don't have any idea why object ids are duplicated. 
Also, after patching https://github.com/ray-project/ray/pull/3771, failure chance reduced to 2% on my machine. But I cannot explain why. 
Still investigating it.

**Comment by raulchen** [comment]
Fixed by #4007
